taichi/transforms/vectorize_half2.cpp

```diff
@@ -0,0 +1,586 @@
+    StoreStmt(old_val, old_val_alloc)
+
+    f16 old_val0 = MatrixPtrStmt(old_val_alloc, 0)
+    f16 old_val1 = MatrixPtrStmt(old_val_alloc, 1)
+
+    alloca_stmt0->replace_all_usages_with(old_val0);
+    alloca_stmt1->replace_all_usages_with(old_val1);
+*/
+
+namespace taichi::lang {
+
+class Half2VectorizationAnalyzer : public BasicStmtVisitor {
+ public:
+  explicit Half2VectorizationAnalyzer(IRNode *node) {
+    node->accept(this);
+  }
+
+  std::unordered_set<AtomicOpStmt *> should_remove;
+  std::map<AtomicOpStmt *, AtomicOpStmt *>
+      should_replace_extern;  // self: other
+  std::map<AtomicOpStmt *, AtomicOpStmt *>
+      should_replace_global_temp;  // self: other
+  std::map<AtomicOpStmt *, AtomicOpStmt *>
+      should_replace_get_ch;  // self: other
+
+  void visit(AtomicOpStmt *stmt) override {
+    // opt-out
+    if (stmt->ret_type != PrimitiveType::f16) {
+      return;
+    }
+
+    if (stmt->op_type != AtomicOpType::add) {
+      return;
+    }
+
+    int stmt_type;  // 0: ExternalPtrStmt, 1: GlobalTemporaryStmt, 2: GetChStmt
+    if (stmt->dest->is<ExternalPtrStmt>() &&
+        stmt->dest->cast<ExternalPtrStmt>()->indices.back()->is<ConstStmt>()) {
+      stmt_type = 0;
+    } else if (stmt->dest->is<GlobalTemporaryStmt>()) {
+      stmt_type = 1;
+    } else if (stmt->dest->is<GetChStmt>()) {
+      stmt_type = 2;
+    } else {
+      return;
+    }
+
+    std::vector<AtomicOpStmt *> atomic_ops_to_remove;
+    for (auto iter : recorded_atomic_ops_) {
+      auto *atomic_op = iter;
+      if (stmt_type == 0) {
+        // [AtomicOpStmt with ExternalPtrStmt]
+        // vectorization patterns:
+        // 1. Same ExternalPtrStmt->base_ptr
+        // 2. Absolute diff of ExternalPtrStmt->offset is 1 (element)
+        auto *self_extern_stmt = stmt->dest->cast<ExternalPtrStmt>();
+        auto *other_extern_stmt = atomic_op->dest->cast<ExternalPtrStmt>();
+
+        if (self_extern_stmt->base_ptr != other_extern_stmt->base_ptr) {
+          continue;
+        }
+
+        std::vector<Stmt *> self_extern_indices = self_extern_stmt->indices;
+        std::vector<Stmt *> other_extern_indices = other_extern_stmt->indices;
+        if (self_extern_indices.size() != other_extern_indices.size()) {
+          continue;
+        }
+
+        if (self_extern_indices.back()->cast<ConstStmt>()->val.val_int32() +
+                other_extern_indices.back()
+                    ->cast<ConstStmt>()
+                    ->val.val_int32() ==
+            1) {
+          // Found pair
+          atomic_ops_to_remove.push_back(iter);
+
+          if (self_extern_indices.back()->cast<ConstStmt>()->val.val_int32() ==
+              0) {
+            should_remove.insert(atomic_op);
+            should_replace_extern[stmt] = atomic_op;
+          } else {
+            should_remove.insert(stmt);
+            should_replace_extern[atomic_op] = stmt;
+          }
+        }
+      }
+
+      if (stmt_type == 1) {
```

Comment:
For global temp, is it possible to pair `vec1.y` and `vec2.x` if they are placed next to each other?